### PR TITLE
Add NodeJS to CBL mariner build containers

### PR DESCRIPTION
**Pull request by mmitche**
Required for aspnetcore

### CODE REVIEW

**Review by akoeplinger — APPROVED** [review]

**Review by ViktorHofer — COMMENTED** [review]
Please double check that this is the right version (nodejs 20). On Ubuntu 22.04 we had to install it from a different registry to get the newer version: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/commit/e9c8822b4a68d9c81a722ab991c93642644fcbcd

**Comment by mmitche** [comment]
It's installing 16.20, so more work is needed here.

**Comment by mmitche** [comment]
Downloading the normal binary layout works fine.

**Review by ViktorHofer — APPROVED** [review]
LGTM if that works

File: src/cbl-mariner/2.0/crossdeps-builder/Dockerfile [hunk 1]
```diff
@@ -26,6 +26,11 @@ RUN wget https://packages.microsoft.com/cbl-mariner/2.0/prod/extended/x86_64/con
     tdnf install -y \
         libbsd-devel
 
+ENV NODE_VERSION 20.11.1
+RUN mkdir /usr/share/node && \
+    curl -sSL https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-x64.tar.xz | tar xf - -J -C /usr/share/node --strip-components=1 && \
```
Comment (akoeplinger) [inline src/cbl-mariner/2.0/crossdeps-builder/Dockerfile:31]: please add checking the sha like we do for llvm-project.src.tar.xz below

File: src/cbl-mariner/2.0/crossdeps-builder/Dockerfile [hunk 2]
```diff
@@ -26,6 +26,11 @@ RUN wget https://packages.microsoft.com/cbl-mariner/2.0/prod/extended/x86_64/con
     tdnf install -y \
         libbsd-devel
 
+ENV NODE_VERSION 20.11.1
```
Comment (akoeplinger) [inline src/cbl-mariner/2.0/crossdeps-builder/Dockerfile:29]:
```suggestion
ENV NODE_VERSION=20.11.1
```

the other syntax is deprecated according to docker docs

File: src/cbl-mariner/2.0/crossdeps-builder/Dockerfile (hunk 1 above)
Comment (mmitche) [reply src/cbl-mariner/2.0/crossdeps-builder/Dockerfile:31]: The process to check the downloaded file is https://github.com/nodejs/node?tab=readme-ov-file#verifying-binaries

Unfortunately, the key import part does not work on mariner right now. See https://github.com/microsoft/azurelinux/issues/3142. There is a [PR open](https://github.com/microsoft/azurelinux/pull/8349) to resolve this, but the change (install gnutls-devel) doesn't appear make a difference. Leaving the verification of SHA256SUM.txt for later then,

**Comment by mmitche** [comment]
I need an expert here. I perhaps wrongly assumed that changing https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/cbl-mariner/2.0/crossdeps-builder/Dockerfile would then cascade to https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/cbl-mariner/2.0/cross/amd64/Dockerfile due to https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/cbl-mariner/2.0/cross/amd64/Dockerfile#L1 but that doesn't appear to be the case.

What am I missing?

**Comment by mmitche** [comment]
Note that mcr.microsoft.com/dotnet-buildtools/prereqs:cbl-mariner-2.0-crossdeps did get node as expected.

**Comment by MichaelSimons** [comment]
The crossdeps-builder images is used to build the rootfs that is then copied in the final cross image.  The key piece you missed is that the cross images are multi-staged images.  see https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/cbl-mariner/2.0/cross/amd64/Dockerfile#L44.

**Comment by MichaelSimons** [comment]
Looks like you probably want to install NodeJS in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/cbl-mariner/2.0/crossdeps/Dockerfile

**Comment by mmitche** [comment]
https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/978

**Pull request merged**